packages/go_router/lib/src/on_enter.dart

```diff
@@ -0,0 +1,307 @@
+// ignore_for_file: use_build_context_synchronously
+
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+import 'configuration.dart';
+import 'information_provider.dart';
+import 'match.dart';
+import 'misc/errors.dart';
+import 'parser.dart';
+import 'router.dart';
+import 'state.dart';
+
+/// The result of an [onEnter] callback.
+///
+/// This sealed class represents the possible outcomes of navigation interception.
+sealed class OnEnterResult {
+  /// Creates an [OnEnterResult].
+  const OnEnterResult();
+}
+
+/// Allows the navigation to proceed.
+final class Allow extends OnEnterResult {
+  /// Creates an [Allow] result.
+  const Allow();
+}
+
+/// Blocks the navigation from proceeding.
+final class Block extends OnEnterResult {
+  /// Creates a [Block] result.
+  const Block();
+}
+
+/// The signature for the top-level [onEnter] callback.
+///
+/// This callback receives the [BuildContext], the current navigation state,
+/// the state being navigated to, and a reference to the [GoRouter] instance.
+/// It returns a [Future<OnEnterResult>] which should resolve to [Allow] if navigation
+/// is allowed, or [Block] to block navigation.
+typedef OnEnter = Future<OnEnterResult> Function(
+  BuildContext context,
+  GoRouterState currentState,
+  GoRouterState nextState,
+  GoRouter goRouter,
+);
+
+/// Handles the top-level [onEnter] callback logic and manages redirection history.
+///
+/// This class encapsulates the logic to execute the top-level [onEnter] callback,
+/// enforce the redirection limit defined in the router configuration, and generate
+/// an error match list when the limit is exceeded. It is used internally by [GoRouter]
+/// during route parsing.
+class OnEnterHandler {
```

Comment:
if possible lets make this private in parser.dart so we don't expose more api surface